File: frontend/src/lib/speaker-samples.ts
/**
 * Speaker sample selection (F1 identification evidence).
 *
 * Groups a meeting's transcript segments by their diarized `speakerId` and picks
 * a few representative lines per speaker so the user has something to READ and
 * HEAR when deciding who a "Speaker N" actually is. Pure functions over the
 * segments already loaded in the transcript view — no backend call needed.
 *
 * No-Fake-State: samples are the real transcribed text at real timestamps. When
 * a speaker has no attributed lines, the result is simply empty.
 */

import type { TranscriptSegmentData } from '@/types';

export interface SpeakerSample {
  /** The transcript segment id (stable key). */
  id: string;
  /** The transcribed text of the line. */
  text: string;
  /** Recording-relative start time in seconds (drives clip playback). */
  startSeconds: number;
}

/** Format a recording-relative second offset as `MM:SS`. */
export function formatClock(seconds: number): string {
  const total = Math.max(0, Math.floor(seconds));
  const minutes = Math.floor(total / 60);
  const secs = total % 60;
  return `${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
}

/**
 * Pick up to `max` representative lines a speaker said. We prefer the most
 * substantive (longest) lines — the best voiceprint evidence — then present
 * them in chronological order so the timestamps read naturally.
 */
export function selectSpeakerSamples(
  segments: TranscriptSegmentData[],
  speakerId: string,
  max = 5,
): SpeakerSample[] {
  const mine = segments.filter(
    (s) => s.speakerId === speakerId && s.text.trim().length > 0,
  );
  return [...mine]
    .sort((a, b) => b.text.trim().length - a.text.trim().length)
    .slice(0, max)
    .sort((a, b) => (a.timestamp ?? 0) - (b.timestamp ?? 0))
    .map((s) => ({
      id: s.id,
      text: s.text.trim(),
      startSeconds: s.timestamp ?? 0,
    }));
}

/** Build a speakerId → representative-samples map for every diarized speaker. */
export function groupSamplesBySpeaker(
  segments: TranscriptSegmentData[],
  max = 5,
): Map<string, SpeakerSample[]> {
  const ids = new Set<string>();
  for (const s of segments) {
    if (s.speakerId) ids.add(s.speakerId);
  }
  const map = new Map<string, SpeakerSample[]>();
  for (const id of ids) {
    map.set(id, selectSpeakerSamples(segments, id, max));
  }
  return map;
}

File: AriKit/Sources/AriViewModels/SpeakerSamples.swift
//
//  SpeakerSamples.swift — speaker sample selection (F1 identification evidence).
//
//  Port of the frozen Rust/React app's `frontend/src/lib/speaker-samples.ts`. Groups a
//  meeting's transcript segments by their diarized `speakerId` and picks a few representative
//  lines per speaker, so the user has something to READ and HEAR when deciding who a
//  "Speaker N" actually is. Pure functions over the transcript rows already loaded by
//  `MeetingDetailViewModel` — no backend call needed.
//
//  No-Fake-State: samples are the real transcribed text at real timestamps. When a speaker has
//  no attributed, timed lines, the result is simply empty.
//
import AriKit
import Foundation

public enum SpeakerSamples {
    /// One representative transcript line, selected as evidence for a speaker's identity.
    public struct SpeakerSample: Identifiable, Equatable, Sendable {
        /// The transcript segment id (stable key).
        public let id: TranscriptID
        /// The trimmed transcribed text of the line.
        public let text: String
        /// Recording-relative start time in seconds (drives clip playback).
        public let startSeconds: Double

        public init(id: TranscriptID, text: String, startSeconds: Double) {
            self.id = id
            self.text = text
            self.startSeconds = startSeconds
        }
    }

    /// Picks up to `max` representative lines a speaker said. Prefers the most substantive
    /// (longest) lines — the best voiceprint evidence — then presents them in chronological
    /// order so the timestamps read naturally. Only rows with a matching `speakerId`,
    /// non-empty trimmed text, AND a known `audioStartTime` (needed to play a clip) qualify.
    public static func select(from transcripts: [Transcript], speakerId: SpeakerID, max: Int = 5) -> [SpeakerSample] {
        let mine = transcripts.filter { row in
            guard row.speakerId == speakerId, row.audioStartTime != nil else { return false }
            return !row.transcript.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
        let longestFirst = mine.sorted { lhs, rhs in
            let lhsLength = lhs.transcript.trimmingCharacters(in: .whitespacesAndNewlines).count
            let rhsLength = rhs.transcript.trimmingCharacters(in: .whitespacesAndNewlines).count
            return lhsLength > rhsLength
        }
        let picked = longestFirst.prefix(max)
        let chronological = picked.sorted { lhs, rhs in
            (lhs.audioStartTime ?? 0) < (rhs.audioStartTime ?? 0)
        }
        return chronological.map { row in
            SpeakerSample(
                id: row.id,
                text: row.transcript.trimmingCharacters(in: .whitespacesAndNewlines),
                startSeconds: row.audioStartTime ?? 0
            )
        }
    }

    /// Builds a speakerId → representative-samples map for every diarized speaker present in
    /// `transcripts`. Rows with no `speakerId` are ignored (not yet diarized).
    public static func group(from transcripts: [Transcript], max: Int = 5) -> [SpeakerID: [SpeakerSample]] {
        var ids: Set<SpeakerID> = []
        for row in transcripts {
            if let speakerId = row.speakerId {
                ids.insert(speakerId)
            }
        }
        var result: [SpeakerID: [SpeakerSample]] = [:]
        for id in ids {
            result[id] = select(from: transcripts, speakerId: id, max: max)
        }
        return result
    }
}
